### PR TITLE
Improve assertThrows, use it in NumericRangeTest

### PR DESCRIPTION
**Pull request by som-snytt**

### CODE REVIEW

File: src/testkit/scala/tools/testkit/AssertUtil.scala [hunk 1]
```diff
@@ -93,11 +93,14 @@ object AssertUtil {
...
 
+  private val Unthrown = new ControlThrowable {}
+
   def assertThrown[T <: Throwable: ClassTag](checker: T => Boolean)(body: => Any): Unit =
     try {
       body
-      fail("Expression did not throw!")
+      throw Unthrown
     } catch {
+      case _: Unthrown.type   => fail("Expression did not throw!")
```
Comment (lrytz) [inline src/testkit/scala/tools/testkit/AssertUtil.scala]:
```suggestion
      case Unthrown => fail("Expression did not throw!")
```

File: test/junit/scala/collection/IteratorTest.scala [hunk 2]
```diff
@@ -660,16 +660,18 @@ class IteratorTest {
...
+
+    def checkHasElement() = assertNotReachable(Option(seq1.get).map(_.apply(1)).orNull, it)(noop)
 
     assert(it.hasNext)
     assertEquals("first", it.next())
 
     // verify that we're in the middle of seq1
-    assertThrows[AssertionError](checkHasElement())
-    assertThrows[AssertionError](check())
+    checkHasElement()
```
Comment (lrytz) [inline test/junit/scala/collection/IteratorTest.scala]: maybe the issue here is that `assertNotReachable` doesn't visit array elements?

**Review by lrytz — COMMENTED** [review]
Thanks for spotting this! Will push a commit with some suggestions.

File: test/junit/scala/collection/IteratorTest.scala (hunk 2 above)
Comment (som-snytt) [reply test/junit/scala/collection/IteratorTest.scala]: I knew there was an issue but didn't have time to look into it.

**Comment by som-snytt** [comment]
Thanks for your generous comment. I'd intended to get back to the problematic test, but I forgot to make it a draft PR.

This is the sort of case where a student says they had a great teacher who made an encouraging remark instead of a mean comment that made them drop out of the program.

**Comment by lrytz** [comment]
👆 this is the overly modest top scala/scala contributor who started off 8 years ago with a deep type checker patch and has not taken a break ever since

**Comment by som-snytt** [comment]
@lrytz For extra credit, I will try to determine if this new test failure under Travis is for real. Apparently, it is a test I contributed, but then they deleted the comment that tells me what it does...
```
scala.concurrent.FutureTest.bug$divissues$hash10513$u0020firstCompletedOf$u0020must$u0020not$u0020leak$u0020references failed: java.lang.AssertionError: Root Future(<not completed>) held reference java.lang.Object@3d785ff2, took 0.011 sec

[error]     at scala.tools.testkit.AssertUtil$.loop$2(AssertUtil.scala:138)

[error]     at scala.tools.testkit.AssertUtil$.$anonfun$assertNotReachable$5(AssertUtil.scala:148)

[error]     at scala.tools.testkit.AssertUtil$.$anonfun$assertNotReachable$5$adapted(AssertUtil.scala:144)
```

**Pull request merged**